sourcetracker/_cli/gibbs.py

```diff
@@ -134,96 +131,69 @@ def gibbs(table_fp, mapping_fp, output_dir, loo, jobs, alpha1, alpha2, beta,
+    # Identify source and sink samples.
+    source_samples = get_samples(sample_metadata, source_sink_column,
+                                 source_column_value)
+    sink_samples = get_samples(sample_metadata, source_sink_column,
+                               sink_column_value)
 
     # If we have no source samples neither normal operation or loo will work.
     # Will also likely get strange errors.
     if len(source_samples) == 0:
-        raise ValueError('Mapping file or biom table passed contain no '
-                         '`source` samples.')
+        raise ValueError('Sample metadata contains no `source` samples under '
+                         'the given `source_sink_column`. Without source '
+                         'samples neither LOO or normal prediction will work.')
 
     # Prepare the 'sources' matrix by collapsing the `source_samples` by their
     # metadata values.
-    sources_envs, sources_data = collapse_sources(source_samples,
-                                                  sample_metadata,
-                                                  source_category_column,
-                                                  biom_table, sort=True)
+    csources = collapse_source_data(sample_metadata, feature_table,
+                                    source_samples, source_category_column,
+                                    'mean')
 
-    # Rarefiy data if requested.
-    sources_data, biom_table = \
-        subsample_sources_sinks(sources_data, sink_samples, biom_table,
-                                source_rarefaction_depth,
-                                sink_rarefaction_depth)
+    # Rarify collapsed source data if requested.
+    if source_rarefaction_depth > 0:
+        csources = subsample_dataframe(csources, source_rarefaction_depth)
 
-    # Build function that require only a single parameter -- sample -- to
-    # enable parallel processing if requested.
-    if loo:
-        f = partial(_cli_loo_runner, source_category=source_category_column,
-                    alpha1=alpha1, alpha2=alpha2, beta=beta,
-                    restarts=restarts, draws_per_restart=draws_per_restart,
-                    burnin=burnin, delay=delay,
-                    sample_metadata=sample_metadata,
-                    sources_data=sources_data, sources_envs=sources_envs,
-                    biom_table=biom_table, output_dir=output_dir)
-        sample_iter = source_samples
-    else:
-        f = partial(_cli_sink_source_prediction_runner, alpha1=alpha1,
-                    alpha2=alpha2, beta=beta, restarts=restarts,
-                    draws_per_restart=draws_per_restart, burnin=burnin,
-                    delay=delay, sources_data=sources_data,
-                    biom_table=biom_table, output_dir=output_dir)
-        sample_iter = sink_samples
+    # Prepare sink data and rarify if requested.
+    sinks = feature_table.loc[sink_samples, :]
+    if sink_rarefaction_depth > 0:
```

Comment:
Different rarefaction depths are important for e.g. seeing how results change as more sequence mass is included. I don't think people will be passing in negative or 0 rarefaction depths, but I will add a check just in case.